### DISCUSSION TITLE
Issue with doorbell camera in HomeKit

### DISCUSSION THREAD

**Discussion by robotDad414** [homebridge-simplisafe3/homebridge-simplisafe3#334] [Q&A]
I'm currently running  v1.10.2 and I noticed that the doorbell camera shows as "No Response" in the Homekit App on iOS. I turned on debugging and opened the app and the error message below is what is getting thrown. If I'm reading it correctly it's just a simple type error and the file path isn't being read in as a string. Is anyone else having the issue? Am I totally misunderstanding the problem? 

[Home Alarm] Unable to spawn ffmpeg process at undefined with error: TypeError: The "file" argument must be of type string. Received undefined
    at new NodeError (node:internal/errors:371:5)
    at validateString (node:internal/validators:120:11)
    at normalizeSpawnArguments (node:child_process:505:3)
    at spawn (node:child_process:693:13)
    at StreamingDelegate.handleStreamRequest (/usr/local/lib/node_modules/homebridge-simplisafe3/lib/streamingDelegate.js:372:48) {
  code: 'ERR_INVALID_ARG_TYPE'
}

**Comment by shamoon** [comment]
Actually I would guess this is the same as #332 which a few people have been experiencing lately. Did you install (or upgrade) the plug-in via the UI or command line? 

Can you try installing like `sudo npm install -g --unsafe-perm homebridge-simplisafe3`

**Reply by shamoon** [reply]
Actually I think there’s a bug in the version of the ffmpeg tool we’re using, I will have to release a new version of the plug-in, will get back to you soon

**Comment by shamoon** [comment]
Ok please try updating to v1.10.3, I believe that will fix this issue, one of the plugin dependencies had a bug in it.

**Reply by robotDad414** [reply]
Just installed the update via the UI and I'm still getting the same error.

[4/14/2022, 6:45:18 PM] [Home Alarm] Unable to spawn ffmpeg process at undefined with error: TypeError: The "file" argument must be of type string. Received undefined
    at new NodeError (node:internal/errors:371:5)
    at validateString (node:internal/validators:120:11)
    at normalizeSpawnArguments (node:child_process:505:3)
    at spawn (node:child_process:693:13)
    at StreamingDelegate.handleStreamRequest (/usr/local/lib/node_modules/homebridge-simplisafe3/lib/streamingDelegate.js:372:48) {
  code: 'ERR_INVALID_ARG_TYPE'
}

**Reply by shamoon** [reply]
Can you remove the plug-in and then reinstall? Make sure to take a backup first and you can tell homebridge not to remove deleted plugins. You don’t need to restart in between

**Reply by robotDad414** [reply]
I gave reinstalling a shot, but still no dice

**Reply by robotDad414** [reply]
It's weird that it's able to take a snapshot, but it cannot connect for streaming.

**Reply by shamoon** [reply]
Thats expected because snapshots dont use ffmpeg.

Hmm, odd, that fixed it for others. can you give me details of what your setup is e.g. OS, install type etc?

This will probably fix things for you: run `node HOMEBRIDGE_DIR/homebridge-simplisafe3/node_modules/ffmpeg-for-homebridge/install.js`. Note that the `HOMEBRIDGE_DIR` varies by install type, often its `/usr/local/lib/node_modules/` but again it depends.

And, if youre feeling generous, I would be curious to see the output of running `npm install -g --unsafe-perm homebridge-simplisafe3`

Basically, after this plugin installs it runs a script from one of the dependencies which is what actually downloads `ffmpeg`, we should see that when you run the above.

Either way, thanks for your patience.

**Reply by robotDad414** [reply]
Shamoon,
 Sorry for the late reply, I tried everything you listed above, but I'm still getting the same error. I'm going to do a completely clean install tonight and try your suggestion above again.

**Reply by robotDad414** [reply]
I hoped that was going to solve my issues but no dice. I'm running on my server on a 2021 Mac Mini, Mac OS Monterey vs 12.3.1. I've installed your plugin via the command line and using the homebridge UI installer.

**Reply by shamoon** [reply]
People on Reddit having the same issue but reinstalling helps. What about when you run the commands i mentioned eg the install.js? The actual output could be helpful to see.

In general you just need to get ffmpeg installed and then you can point the plug-in at it. 

Btw is that an Apple silicon Mac? Should be supported https://github.com/homebridge/ffmpeg-for-homebridge/issues/30

**Reply by shamoon** [reply]
Btw, I created a beta `1.10.4-beta.1` that you may want to try, it is just a cleaned up version of the dependencies, but not sure will make a difference since Im still not 100% clear on the issue. Again, output might be helpful, see my previous comment.

**Reply by snoslicer8** [reply]
How can we install the beta over the released version? I am having this same issue and none of the solutions above worked for me.

**Reply by snoslicer8** [reply]
Never mind. I installed the beta through the Homebridge UI, and still the camera is “not responding.” With Homebridge error log of:

4/17/2022, 3:31:55 PM] [Home Alarm] Unable to spawn ffmpeg process at undefined with error: TypeError: The "file" argument must be of type string. Received undefined
    at new NodeError (node:internal/errors:371:5)
    at validateString (node:internal/validators:119:11)
    at normalizeSpawnArguments (node:child_process:503:3)
    at spawn (node:child_process:691:13)
    at StreamingDelegate.handleStreamRequest (/usr/local/lib/node_modules/homebridge-simplisafe3/lib/streamingDelegate.js:372:48) {
  code: 'ERR_INVALID_ARG_TYPE'

**Reply by shamoon** [reply]
did you try running https://github.com/homebridge-simplisafe3/homebridge-simplisafe3/discussions/334#discussioncomment-2570598 that should install ffmpeg directly

**Reply by shamoon** [reply]
Also @snoslicer8 what platform are you using and how did you install?

**Reply by snoslicer8** [reply]
Yes, I did the steps in #334 and it said it installed successfully. But still no dice. I am running on an M1 Mac Mini, latest OS, installed via Homebridge UI.

**Reply by snoslicer8** [reply]
Notably, the video used to work with the plug-in, but stopped sometime over the last few weeks - I don’t know exactly when, I’m sorry.

**Reply by shamoon** [reply]
Hmm, youre both on Apple Silicon macs.

@snoslicer8 so you ran the install.js command, and it worked? And then do you have an actual binary at `/usr/local/lib/node_modules/homebridge-simplisafe3/node_modules/ffmpeg-for-homebridge/ffmpeg` ?

**Comment by snoslicer8** [comment]
This is the output I get:
![89BD23E3-E34C-471F-B51F-0C1F47C305F8](https://user-images.githubusercontent.com/650125/163734043-b3050f32-16f6-4a37-b5f9-b58899b481ae.png)

**Reply by shamoon** [reply]
Thank you, this is what I was looking for.

```
node install.js downloading ffmpeg from https://github.com/homebridge/ffmpeg-for-homebridge/releases/latest/download/ffmpeg -darwin-x86_64. tar. gz
Download Complete yld(57041]: Library not loaded: /usr/local/opt/libx11/lib/libX11.6.dylib 
Referenced from: /Users/blandgren/.npm/_ffmpeg_for_homebridge/ffmpeg 
Reason: tried: '/usr/local/opt/libx11/lib/libX11.6.dylib' (no such file), '/usr/local/lib/libX11.6.dylib' (no such file), '/usr/lib/libX11.6.dylib' (no such file)
```

Its downloading the x86 package. Can you try

`
sudo curl -Lf# https://github.com/homebridge/ffmpeg-for-homebridge/releases/download/v0.1.0/ffmpeg-darwin-arm64.tar.gz | sudo tar xzfm - -C / --no-same-owner
`

**Reply by snoslicer8** [reply]
That command ran fine, but didn’t give me any feedback in terminal. I restarted Homebridge, and still getting the same error in Homebridge logs when I try to view the camera in the Home app.

**Reply by shamoon** [reply]
Ive opened an issue with the dependency plugin, its clearly the issue. I had thought they support Apple Silicon so either they dont or something broke. Will follow-up

Edit: you already had a binary installed, just the wrong one. If you want you can try to debug a little, remove the incorrect one and try to install the new one, but yea, you can also just wait till I hear back about the dependency.

**Reply by snoslicer8** [reply]
This is the content of that directory:
README.md	index.d.ts	index.js	install.js	package.json

**Reply by snoslicer8** [reply]
> Ive opened an issue with the dependency plugin, its clearly the issue. I had thought they support Apple Silicon so either they dont or something broke. Will follow-up
> 
> Edit: you already had a binary installed, just the wrong one. If you want you can try to debug a little, remove the incorrect one and try to install the new one, but yea, you can also just wait till I hear back about the dependency.

Thanks. As I said before, it did originally work when I installed this plugin just about a month ago. Same machine and OS setup, so I think something broke recently in the plug-in or dependency, as you said.

**Comment by robotDad414** [comment]
I figured it out . So the issue is with X11 . FFmpeg is searching for the X11 installation at a couple of different directories . Specifically it's look for libX11.6.dylib. I used  sudo find / -name libX11.6.dylib and found the file and then I just made a direct copy into /usr/local/lib/ and then reinstalled ffmpeg and now it's working .

**Reply by robotDad414** [reply]
Also sorry I was with the fam all day and didn't have access to my server.

**Reply by shamoon** [reply]
Well, one might say "we" figured it out =) , but glad to hear.

**Reply by robotDad414** [reply]
Yes , sorry shamoon , it was definitely a we ( I definitely couldn't put together any of these plugins) . This is the error that it was throwing after the install of ffmpeg. 

Pleximus:ffmpeg-for-homebridge pleximusbeyond$ node install.js
Downloading ffmpeg from https://github.com/homebridge/ffmpeg-for-homebridge/releases/latest/download/ffmpeg-darwin-x86_64.tar.gz
100% - Download Complete
dyld[86648]: Library not loaded: /usr/local/opt/libx11/lib/libX11.6.dylib
  Referenced from: /Users/pleximusbeyond/.npm/_ffmpeg_for_homebridge/ffmpeg
  Reason: tried: '/usr/local/opt/libx11/lib/libX11.6.dylib' (no such file), '/usr/local/lib/libX11.6.dylib' (no such file), '/usr/lib/libX11.6.dylib' (no such file)

The homebridge plugin has been installed, however you may need to install ffmpeg separately.

**Reply by snoslicer8** [reply]
> Well, one might say "we" figured it out =) , but glad to hear.

Thank you gents! I found a copy of that file somewhere on the server and copied it to the referenced location, rebooted Homebridge, and the camera works in Home again now! Well done, guys.

**Reply by shamoon** [reply]
Ha yes, though there is still an issue with that package we will have to wait on but glad to have a solution in the meantime. Thanks for all your help investigating
